File: lib/biz/spec-iot/src/main/java/site/addzero/biz/spec/iot/IotThingRef.java
package site.addzero.biz.spec.iot;

import java.util.Objects;

/**
 * Generic thing reference used by schema and telemetry APIs.
 */
public final class IotThingRef {

    private final String kind;
    private final String id;

    private IotThingRef(String kind, String id) {
        this.kind = requireText(kind, "kind");
        this.id = requireText(id, "id");
    }

    public static IotThingRef of(String kind, String id) {
        return new IotThingRef(kind, id);
    }

    public String getKind() {
        return kind;
    }

    public String getId() {
        return id;
    }

    private static String requireText(String value, String field) {
        if (value == null || value.trim().isEmpty()) {
            throw new IllegalArgumentException(field + " must not be blank");
        }
        return value.trim();
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof IotThingRef)) {
            return false;
        }
        IotThingRef that = (IotThingRef) other;
        return kind.equals(that.kind) && id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(kind, id);
    }

    @Override
    public String toString() {
        return kind + ":" + id;
    }
}

File: lib/biz/spec-iot/src/main/kotlin/site/addzero/biz/spec/iot/IotThingRef.kt
package site.addzero.biz.spec.iot

/**
 * Generic thing reference used by schema and telemetry APIs.
 */
class IotThingRef private constructor(
    val kind: String,
    val id: String,
) {

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is IotThingRef) {
            return false
        }
        return kind == other.kind && id == other.id
    }

    override fun hashCode(): Int {
        var result = kind.hashCode()
        result = 31 * result + id.hashCode()
        return result
    }

    override fun toString(): String {
        return "$kind:$id"
    }

    companion object {
        @JvmStatic
        fun of(kind: String?, id: String?): IotThingRef {
            return IotThingRef(
                kind = requireText(kind, "kind"),
                id = requireText(id, "id"),
            )
        }
    }
}
